### PR TITLE
Migration script

### PR DESCRIPTION
**Pull request by shankgan**

### CODE REVIEW

**Comment by shankgan** [comment]
Preliminary. Do not merge

File: scripts/migration/migrate-14-managed
Comment (shankgan) [inline scripts/migration/migrate-14-managed:243]: doesn't work...will be rolled back in the next commit

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:1]: nit: executables shouldn't have .sh extensions

Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:1]: Use `#!/usr/bin/env bash` instead for compatibility

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:33]: Please surround $() with double quotes

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:181]: I don't think I've ever once installed `istioctl`. Can we just have this tool download and extract it if necessary?

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:190]: nit: popd is unnecessary if you're going to `die` right afterwards

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:228]: Can we check here and do some better error reporting if the namespace still exists?

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:222]: Can we just get the namespaces with the injection label?

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:259]: Is this possible to roll back?

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:256]: `replace_galley_webhook` seems like a better name

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:457]: unnecessary semicolon

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:479]: Get rid of extra period at the end. Also, the error messages aren't consistent with punctuation or capitalization.

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [inline scripts/migration/migrate-addon.sh:587]: Let's make it more obvious that if you leave out `--command` it runs all of them.

**Review by zerobfd — CHANGES_REQUESTED** [review]
Overall looks OK. My biggest request is to add TODO comments to places where we know there's work, so it's more apparent to reviewers.

Also, unless you source a shell file (which is mostly a bad idea unless you know what you're doing) then it can't affect your environment, so all of the pushd/popd is unnecessary.

File: scripts/migration/migrate-addon.sh
Comment (shankgan) [reply scripts/migration/migrate-addon.sh:181]: The tool isn't meant to run in isolation. Since we also expect the Managed control-plane to be installed prior to the migration tool, that step should end up installing istioctl. The docs will reflect this.

File: scripts/migration/migrate-addon.sh
Comment (shankgan) [reply scripts/migration/migrate-addon.sh:256]: This actually is just disabling the galley webook. Will change the header message.

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [reply scripts/migration/migrate-addon.sh]: We can rely on the docs to guide users to download `istioctl`, but "installing" it (e.g. putting it in $PATH) isn't something that's normally done (to my knowledge). We should support a way to either specify the `istioctl` binary location or put it in a consistent location ourselves.

Comment (shankgan) [reply scripts/migration/migrate-addon.sh]: To be more specific: The first step in the docs is to use the scriptaro installer to download the appropriate version of istioctl. The location of the binary is one of the arguments to the installer script. We could add the same argument here to this script and add logic to find istioctl there - does that work?

File: scripts/migration/migrate-addon.sh
Comment (shankgan) [reply scripts/migration/migrate-addon.sh]: The tool now requires an explicit command for it to "run everything". This command is hidden from usages as it is not meant to be used in interactive mode

File: scripts/migration/migrate-addon.sh
Comment (shankgan) [reply scripts/migration/migrate-addon.sh]: we now store the configuration before deleting it..the rollback can rely on that if needed

File: scripts/migration/migrate-addon.sh
Comment (zerobfd) [reply scripts/migration/migrate-addon.sh]: Sure. Or we could make it `./istioctl` and have users `cd` into the output directory that `install_asm` makes, since it puts a copy of `istioctl` there.

**Review by zerobfd — APPROVED** [review]
lgtm after you decide what to do with the `istioctl` thing. If you haven't made a bug/issue already then make sure you have some way to remember to follow up on the TODOs and testing.

**Pull request merged**
